co2calculator/calculate.py

```diff
@@ -453,208 +464,107 @@ def calc_co2_ferry(
-    :type transportation_mode: str
-    :type distance: float
-    :type size: str
-    :type fuel_type: str
-    :type occupancy: int
-    :type seating: str
-    :type passengers: int
-    :type roundtrip: bool
-    :return:    Emissions of the business trip in co2 equivalents,
-                Distance of the business trip,
-                Range category of the business trip [very short haul, short haul, medium haul, long haul]
-                Range description (i.e., what range of distances does to category correspond to)
-    :rtype: tuple[float, float, str, str]
-    """
-    if distance is None and (start is None or destination is None):
-        assert ValueError("Either start and destination or distance must be provided.")
-    elif distance is not None and (start is not None or destination is not None):
-        warnings.warn(
-            "Both distance and start/stop location were provided. "
-            "Only distance will be used for emission calculation."
-        )
-        stops = None
-    elif start is None and destination is None and distance is not None:
-        stops = None
-    elif start is not None and destination is not None and distance is None:
-        # check if stops are provided in the right form
-        if transportation_mode == "car" and (
-            type(start) != str or type(destination) != str
-        ):
-            raise ValueError(
-                "Wrong data type for start and destination."
-                "Please provide a three letter IATA code for train stations."
-            )
-        elif transportation_mode != "car" and (
-            type(start) != dict or type(destination) != dict
-        ):
-            raise ValueError(
-                "Wrong data type for start and destination."
-                "Please provide a dictionary."
+) -> Tuple[float, float, str, str]:
+    """Interface to calculate business trips"""
+    # TODO: Validate that type switch for `transportation_mode` is not a
+    # breaking change
+
+    # All co2 emissions are calculated for given distance. Hence, let's check
+    # if distance was parsed. If not, calculate that first.
+    # NOTE: Imho (Jakob): this should be taken care of by the caller (backend)
+
+    if not distance:
+        if start is None or destination is None:
+            raise InputMissing(
+                "for location based calculations: start and destination must not be None"
             )
 
-        stops = [start, destination]
-    if transportation_mode == "car":
-        emissions, dist = calc_co2_car(
-            distance=distance,
-            stops=stops,
-            passengers=passengers,
-            size=size,
-            fuel_type=fuel_type,
+        start_loc = Stop(**start)
+        dest_loc = Stop(**destination)
+        stops = StopList.parse_obj([start_loc, dest_loc])
+        dist_request = DistanceRequest(
+            stops=stops, transportation_mode=transportation_mode
         )
```

Comment:
```python
dist_request = DistanceRequest(
    stops=[start_loc, dest_loc], 
    transportation_mode=transportation_mode
)
```

.. should work too, if `DistanceRequest.stops: List[Stop]`

Give it a try. You can delete the line above if it works!

Reply:
Or even a little bit more fancy:

```python
dist_request = DistanceRequest(
    stops=[Stop(**s) for s in [start, destination]], 
    transportation_mode=transportation_mode
)
```

.. and delete the three lines above!